base/inference.jl

```diff
@@ -3197,14 +3204,277 @@ function occurs_outside_getfield(linfo::LambdaInfo, e::ANY, sym::ANY,
     return false
 end
 
-# removes inbounds metadata if we never encounter an inbounds=true or
-# boundscheck context in the method body
-function inbounds_meta_elim_pass!(code::Array{Any,1})
-    if findfirst(x -> isa(x, Expr) &&
-                      ((x.head === :inbounds && x.args[1] === true) || x.head === :boundscheck),
-                 code) == 0
-        filter!(x -> !(isa(x, Expr) && x.head === :inbounds), code)
+function void_use_elim_pass!(linfo::LambdaInfo, sv)
+    # Remove top level SSAValue and slots that is `!usedUndef`.
+    # Also remove some `nothing` while we are at it....
+    not_void_use = function (ex::ANY)
+        if isa(ex, SSAValue)
+            # Explicitly listed here for clarity
+            return false
+        elseif isa(ex, Slot)
+            return linfo.slotflags[(ex::Slot).id] & Slot_UsedUndef != 0
+        elseif (isa(ex, Expr) || isa(ex, GotoNode) || isa(ex, LineNumberNode) ||
```

Comment:
Sure and done.

I somehow thought it was already handled in type inference but seems that I was testing the wrong thing....